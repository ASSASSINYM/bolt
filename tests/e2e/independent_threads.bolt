class Foo  {
  capability linear Bar;
  var int f : Bar;
  const int g : Bar;  
  const int h : Bar;

}
function int f (int x ){ x}
void main(){
  let x = new Foo(f:5);
  finish {
    async{
      f(5)
    }
    async{
      let w = new Foo(g:5);
      w.f := 5
    }
    printf("Value of x.f: %d\n", x.f)
  }
}
