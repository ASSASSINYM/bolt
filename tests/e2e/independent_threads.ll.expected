; ModuleID = 'Module'
source_filename = "Module"
target triple = "x86_64-apple-darwin18.7.0"

%struct._opaque_pthread_t = type opaque
%Foo = type { i32, i32, i32 }

@0 = private unnamed_addr constant [19 x i8] c"Value of x.f: %d\5Cn\00", align 1

declare i32 @printf(i8*, ...)

declare i32 @pthread_create(%struct._opaque_pthread_t*, i8*, i8* (i8*)*, i8*)

declare i32 @pthread_join(%struct._opaque_pthread_t*, i8**)

define i32 @f(i32) {
entry:
  %x = alloca i32
  store i32 %0, i32* %x
  %1 = load i32, i32* %x
  ret i32 %1
}

define i32 @main() {
entry:
  %_var_x0 = alloca %Foo
  %Foo = alloca %Foo
  %0 = getelementptr inbounds %Foo, %Foo* %Foo, i32 0, i32 0
  store i32 5, i32* %0
  %1 = load %Foo, %Foo* %Foo
  store %Foo %1, %Foo* %_var_x0
  %pthread = alloca %struct._opaque_pthread_t
  %2 = call i32 @pthread_create(%struct._opaque_pthread_t* %pthread, i8* null, i8* (i8*)* @_async0, i8* null)
  %pthread1 = alloca %struct._opaque_pthread_t
  %3 = call i32 @pthread_create(%struct._opaque_pthread_t* %pthread1, i8* null, i8* (i8*)* @_async1, i8* null)
  %4 = getelementptr inbounds %Foo, %Foo* %_var_x0, i32 0, i32 0
  %5 = load i32, i32* %4
  %6 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([19 x i8], [19 x i8]* @0, i32 0, i32 0), i32 %5)
  %7 = call i32 @pthread_join(%struct._opaque_pthread_t* %pthread, i8** null)
  %8 = call i32 @pthread_join(%struct._opaque_pthread_t* %pthread1, i8** null)
  ret i32 0
}

define i8* @_async0(i8*) {
entry:
  %1 = call i32 @f(i32 5)
  ret i8* null
}

define i8* @_async1(i8*) {
entry:
  %_var_w0 = alloca %Foo
  %Foo = alloca %Foo
  %1 = getelementptr inbounds %Foo, %Foo* %Foo, i32 0, i32 1
  store i32 5, i32* %1
  %2 = load %Foo, %Foo* %Foo
  store %Foo %2, %Foo* %_var_w0
  %3 = getelementptr inbounds %Foo, %Foo* %_var_w0, i32 0, i32 0
  store i32 5, i32* %3
  ret i8* null
}
